javascript/ql/src/semmle/javascript/frameworks/EventEmitter.qll

```diff
@@ -0,0 +1,167 @@
+import javascript
+
+module EventEmitter {
+  /** Gets the name of a method on `EventEmitter` that returns `this`. */
+  string chainableMethod() {
+    result = "off" or
+    result = "removeAllListeners" or
+    result = "removeListener" or
+    result = "setMaxListeners" or
+    result = on()
+  }
+
+  /** Gets the name of a method on `EventEmitter` that registers an event handler. */
+  string on() {
+    result = "addListener" or
+    result = "on" or
+    result = "once" or
+    result = "prependListener" or
+    result = "prependOnceListener"
+  }
+  
+  /**
+   * An instanceof of the NodeJS EventEmitter class. 
+   * Extend this class to mark something as being instanceof the EventEmitter class. 
+   */
+  abstract class EventEmitter extends DataFlow::Node {
+    /**
+     * Get a method name that returns `this` on this type of emitter.
+     */
+    string getAChainableMethod() { result = EventEmitter::chainableMethod() }
+
+    private DataFlow::SourceNode ref(DataFlow::TypeTracker t) {
+      t.start() and result = this
+      or
+      exists(DataFlow::TypeTracker t2, DataFlow::SourceNode pred | pred = ref(t2) |
+        result = pred.track(t2, t)
+        or
+        // invocation of a chainable method
+        exists(DataFlow::MethodCallNode mcn |
+          mcn = pred.getAMethodCall(this.getAChainableMethod()) and
+          // exclude getter versions
+          exists(mcn.getAnArgument()) and
+          result = mcn and
+          t = t2.continue()
+        )
+      )
+    }
+
+    /**
+     * Get a reference through type-tracking to this EventEmitter. 
+     * The type-tracking tracks through chainable methods.
+     */
+    DataFlow::SourceNode ref() { result = ref(DataFlow::TypeTracker::end()) }
+  }
+
+  /**
+   * A registration of an event handler on a particular EventEmitter.
+   */
+  abstract class EventRegistration extends DataFlow::Node {
+    EventEmitter emitter;
+
+    /** Gets the EventEmitter that the event handler is registered on. */
+    final EventEmitter getEmitter() {
+      result = emitter
+    }
+
+    /** Gets the name of the channel if possible. */
+    abstract string getChannel();
+
+    /** Gets the `i`th parameter in the callback registered as the event handler. */
+    abstract DataFlow::Node getCallbackParameter(int i);
```

Comment:
Could we rename this to something more general? Callbacks are the mechanism by which data is passed to the event handler, but perhaps we could make the name about the concept rather than the mechanism?

Reply:
I've renamed to `getEventHandlerParameter`.